blog/zh-cn/dubbo-heartbeat-design.md

```diff
@@ -0,0 +1,386 @@
+        public void callback(Future future) {
+            RPCResult result = future.get();
+            //超时 或者 写失败
+            if (result.isError()) {
+                channel.addFailedHeartBeatTimes();
+                if (channel.getFailedHeartBeatTimes() >= channel.getMaxHeartBeatFailedTimes()) {
+                    channel.reconnect();
+                }
+            } else {
+                channel.clearHeartBeatFailedTimes();
+            }
+        }
+    });
+}
+```
+
+行为并不复杂，构造一个心跳包发送到服务端，接受响应结果
+
+- 响应成功，清空请求失败标记
+- 响应失败，心跳失败标记+1，如果超过配置的失败次数，则重新连接
+
+> 不仅仅是心跳，普通请求返回成功响应时也会清空标记
+
+#### 4.4 空闲超时逻辑 — 服务端
+
+```java
+@Override
+public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+    if (evt instanceof IdleStateEvent) {
+        channel.close();
+    } else {
+        super.userEventTriggered(ctx, evt);
+    }
+}
+```
+
+服务端处理空闲连接的方式非常简单粗暴，直接关闭连接。
+
+#### 4.5 改进方案心跳总结
+
+1. 为什么客户端和服务端配置的超时时间不一致？
+
+   因为客户端有重试逻辑，不断发送心跳失败 n 次之后，才认为是连接断开；而服务端是直接断开，留给服务端时间得长一点。60 * 3 < 200 还说明了一个问题，双方都拥有断开连接的能力，但连接的创建是由客户端主动发起的，那么客户端也更有权利去主动断开连接。
+
+2. 为什么客户端检测的是读超时，而服务端检测的是读写超时？
+
+   这其实是一个心跳的共识了，仔细思考一下，定时逻辑是由客户端发起的，所以整个链路中不通的情况只有可能是：服务端接收，服务端发送，客户端接收。也就是说，只有客户端的 pong，服务端的 ping，pong 的检测是有意义的。
+
+> 主动追求别人的是你，主动说分手的也是你。
+
+利用 `IdleStateHandler` 实现心跳机制可以说是十分优雅的，借助 Netty 提供的空闲检测机制，利用客户端维护单向心跳，在收到 3 次心跳失败响应之后，客户端断开连接，交由异步线程重连，本质还是表现为客户端重连。服务端在连接空闲较长时间后，主动断开连接，以避免无谓的资源浪费。
+
+### 5 心跳设计方案对比
+
+|                      |                        Dubbo 现有方案                        |                    Dubbo 改进方案                    |
+| :------------------: | :----------------------------------------------------------: | :--------------------------------------------------: |
+|     **主体设计**     |                        开启两个定时器                        |       借助 IdleStateHandler，底层使用 schedule        |
+|     **心跳方向**     |                             双向                             |               单向（客户端 -> 服务端）               |
+| **心跳失败判定方式** | 心跳成功更新标记，借助定时器定时扫描标记，如果超过心跳超时周期未更新标记，认为心跳失败。 | 通过判断心跳响应是否失败，超过失败次数，认为心跳失败 |
+|      **扩展性**      | Dubbo 存在 mina，grizzy 等其他通信层实现，自定义定时器很容易适配多种扩展 |         多通信层各自实现心跳，不做心跳的抽象         |
+|      **设计性**      |          编码复杂度高，代码量大，方案复杂，不易维护          |                 编码量小，可维护性强                 |
+
+私下请教过**美团点评的长连接负责人：俞超（闪电侠）**，美点使用的心跳方案和 Dubbo 改进方案几乎一致，可以说该方案是标准实现了。
+
+### 6 Dubbo 实际改动点建议
+
+鉴于 Dubbo 存在一些其他通信层的实现，所以可以保留现有的定时发送心跳的逻辑。
+
+- **建议改动点一：**
+
+双向心跳的设计是不必要的，兼容现有的逻辑，可以让客户端在连接空闲时发送单向心跳，服务端定时检测连接可用性。定时时间尽量保证：客户端超时时间 * 3 ≈ 服务端超时时间
+
+- **建议改动点二：**
+
+去除处理重连和断连的定时任务，Dubbo 可以判断心跳请求是否响应失败，可以借鉴改进方案的设计，在连接级别维护一个心跳失败次数的标记，任意响应成功，清除标记；连续心跳失败 n 次，客户端发起重连。这样可以减少一个不必要的定时器，任何轮询的方式，都是不优雅的。
+
+最后再聊聊可扩展性这个话题。其实我是建议把定时器交给更加底层的 Netty 去做，也就是完全使用 `IdleStateHandler` ，其他通信层组件各自实现自己的空闲检测逻辑，但是 Dubbo 中 mina，grizzy 的兼容问题囿住了我的拳脚，但试问一下，如今的 2019 年，又有多少人在使用 mina 和 grizzy？因为一些不太可能用的特性，而限制了主流用法的优化，这肯定不是什么好事。抽象，功能，可扩展性并不是越多越好，开源产品的人力资源是有限的，框架使用者的理解能力也是有限的，能解决大多数人问题的设计，才是好的设计。哎，谁让我不会 mina，grizzy，还懒得去学呢[摊手]。
```

Comment:
+1 for the last part. I am thinking of dropping the support of mina in Dubbo 3.0